### PR TITLE
Adds ZScore Normalization Technique

### PR DESCRIPTION
**Pull request by owaiskazi19**
### Description
Adds ZScore Normalization Technique

### Related Issues
Resolves https://github.com/opensearch-project/neural-search/issues/376 and https://github.com/opensearch-project/neural-search/issues/1209

### Check List
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/neural-search/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+        return normalizedScore <= 0.0f ? MIN_SCORE : normalizedScore;
+    }
+
+    /**
+     * Result class to hold mean and std dev scores for each sub query
+     */
+    @AllArgsConstructor
+    @Getter
+    private class ZScores {
```
Comment (navneet1v) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:230]: with JDK 21 you can now use record key word with classes like this. Try it out. Ref: https://www.baeldung.com/java-record-keyword

File: src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java [hunk 2]
```diff
@@ -50,8 +51,9 @@ public SearchPhaseResultsProcessor create(
     ) throws Exception {
         Map<String, Object> normalizationClause = readOptionalMap(NormalizationProcessor.TYPE, tag, config, NORMALIZATION_CLAUSE);
         ScoreNormalizationTechnique normalizationTechnique = ScoreNormalizationFactory.DEFAULT_METHOD;
+        String normalizationTechniqueName = MinMaxScoreNormalizationTechnique.TECHNIQUE_NAME;
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java:54]: Why we take a default value here?

File: src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java [hunk 3]
```diff
@@ -73,6 +75,11 @@ public SearchPhaseResultsProcessor create(
                 TECHNIQUE,
                 ArithmeticMeanScoreCombinationTechnique.TECHNIQUE_NAME
             );
+            // case when technique is z score and combination is not arithmetic mean
+            if (normalizationTechniqueName.equals(ZScoreNormalizationTechnique.TECHNIQUE_NAME)
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java:79]: Can you extract this validation out of this method to a separate one? Also add why we are blocking these techniques?

File: src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java [hunk 4]
```diff
@@ -73,6 +75,11 @@ public SearchPhaseResultsProcessor create(
...
+                && !combinationTechnique.equals(ArithmeticMeanScoreCombinationTechnique.TECHNIQUE_NAME)) {
```
[the remark this answers is not in the document]
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java:80]: +1

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+import org.opensearch.neuralsearch.processor.explain.ExplainableTechnique;
+import org.opensearch.neuralsearch.processor.explain.ExplanationDetails;
+
+import java.util.List;
+import java.util.Objects;
+
+import static org.opensearch.neuralsearch.processor.explain.ExplanationUtils.getDocIdAtQueryForNormalization;
+
+@ToString(onlyExplicitlyIncluded = true)
+public class ZScoreNormalizationTechnique implements ScoreNormalizationTechnique, ExplainableTechnique {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:34]: Please Add javadoc of the class.

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    static private float sumScoreDocsArray(final ScoreDoc[] scoreDocs) {
+        float sum = 0;
+        for (ScoreDoc scoreDoc : scoreDocs) {
+            sum += scoreDoc.score;
+        }
+
+        return sum;
+    }
+
+    private ZScores getZScoreResults(final List<CompoundTopDocs> queryTopDocs) {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:200]: Could you add a logic of determining Zscore in comments?

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+                meanPerSubQuery[i] = 0;
+            } else {
+                meanPerSubQuery[i] = sumPerSubquery[i] / elementsPerSubquery[i];
+            }
+        }
+
+        return meanPerSubQuery;
+    }
+
+    static private float[] findStdPerSubquery(
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:157]: Does std stands for standard deviation? If yes could you rename this method more appropriate ?

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+            int subQueryIndex = 0;
+            for (TopDocs topDocs : topDocsPerSubQuery) {
+                numberOfElementsPerSubQuery[subQueryIndex++] += topDocs.totalHits.value();
+            }
+        }
+
+        return numberOfElementsPerSubQuery;
+    }
+
+    static private float[] findMeanPerSubquery(final float[] sumPerSubquery, final long[] elementsPerSubquery) {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:144]:
```suggestion
    static private float[] calculateMeanPerSubquery(final float[] sumPerSubquery, final long[] elementsPerSubquery) {
```

**Review by vibrantvarun — COMMENTED** [review]
Completed 1st round of review

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+        return meanPerSubQuery;
+    }
+
+    static private float[] findStdPerSubquery(
+        final List<CompoundTopDocs> queryTopDocs,
+        final float[] meanPerSubQuery,
+        final long[] elementsPerSubquery,
+        final int numOfScores
+    ) {
+        final double[] deltaSumPerSubquery = new double[numOfScores];
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:163]: any reason why we're doing this manually instead of using function from some library, e.g. DescriptiveStatistics from apache-commons math? Reason being - if that's a well-known library I do have more confidence in it for the edge case scenarios.

File: src/test/java/org/opensearch/neuralsearch/processor/ScoreNormalizationIT.java [hunk 10]
```diff
@@ -156,7 +160,7 @@ public void testL2Norm_whenOneShardAndQueryMatches_thenSuccessful() {
      *         {
      *             "normalization-processor": {
      *                 "normalization": {
-     *                     "technique": "l2"
+     *                     "technique": "min_max"
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/neuralsearch/processor/ScoreNormalizationIT.java:163]: why this change? looks like test is for l2

Comment (owaiskazi19) [reply src/test/java/org/opensearch/neuralsearch/processor/ScoreNormalizationIT.java:163]: It's for [min max](https://github.com/opensearch-project/neural-search/pull/1224/files#diff-03f340a64b84233c32a824c2ce968e980a14d1188724571f4b9108dadbc6ad8cR174)

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java (hunk 9 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java]: This was a really good suggestion. Used `DescriptiveStatistics` to calculate mean, sd, min and max scores

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+import java.util.Objects;
+
+import static org.opensearch.neuralsearch.processor.explain.ExplanationUtils.getDocIdAtQueryForNormalization;
+
+@ToString(onlyExplicitlyIncluded = true)
+public class ZScoreNormalizationTechnique implements ScoreNormalizationTechnique, ExplainableTechnique {
+    @ToString.Include
+    public static final String TECHNIQUE_NAME = "z_score";
+    private static final float SINGLE_RESULT_SCORE = 1.0f;
+    private static final float MIN_SCORE = 0.001f;
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:43]: Computed min and max per sub query

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+import java.util.List;
+import java.util.Objects;
+
+import static org.opensearch.neuralsearch.processor.explain.ExplanationUtils.getDocIdAtQueryForNormalization;
+
+@ToString(onlyExplicitlyIncluded = true)
+public class ZScoreNormalizationTechnique implements ScoreNormalizationTechnique, ExplainableTechnique {
+    @ToString.Include
+    public static final String TECHNIQUE_NAME = "z_score";
+    private static final float SINGLE_RESULT_SCORE = 1.0f;
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:42]: returning max score now when mean is equal to score

File: src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java (hunk 2 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java]: We assigned a default value but with the other logic change. Reverted this.

File: src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java]: Took the suggestion. Added a `validateCombinationTechnique` method for all the normalization techniques

File: src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java [hunk 13]
```diff
@@ -73,6 +75,11 @@ public SearchPhaseResultsProcessor create(
...
+                throw new IllegalArgumentException("Z Score supports only arithmetic_mean combination technique");
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java]: From the bechnmarking z score didn't produce good results for GP and HP. In the design meeting a suggestion was made to only support AP for z score.

File: src/test/java/org/opensearch/neuralsearch/processor/ScoreNormalizationIT.java [hunk 14]
```diff
@@ -331,7 +331,7 @@ public void testZScoreNorm_whenOneShardAndQueryMatches_thenSuccessful() {
         org.hamcrest.MatcherAssert.assertThat(
             exceptionWithGP.getMessage(),
             allOf(
-                containsString("Z Score supports only arithmetic_mean combination technique"),
+                containsString("Z Score does not support geometric mean combination technique\""),
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/neuralsearch/processor/ScoreNormalizationIT.java:334]:
```suggestion
                containsString("Z Score does not support geometric mean combination technique"),
```

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationTechnique.java [hunk 15]
```diff
@@ -20,4 +21,5 @@ public interface ScoreNormalizationTechnique {
      */
     void normalize(final NormalizeScoresDTO normalizeScoresDTO);
 
+    void validateCombinationTechnique(final ScoreCombinationTechnique combinationTechnique) throws IllegalArgumentException;
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationTechnique.java:24]: nit: in future we can create a dto class for validation context, single instance of CombinationTechnique is this scope

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 16]
```diff
@@ -52,12 +72,36 @@ public void normalize(NormalizeScoresDTO normalizeScoresDTO) {
...
+                        zscores.maxPerSubQuery[j],
+                        zscores.minPerSubQuery[j]
+                    );
                 }
             }
         }
     }
 
+    @Override
+    public void validateCombinationTechnique(ScoreCombinationTechnique combinationTechnique) throws IllegalArgumentException {
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:88]: this approach is good, but I think we can do it more elegantly. How about validating it as part of the norm technique constructor? We can switch from Function to BiFunction, or create a new dto class  and leave Function, both approaches are fine. Then we pass combination technique to norm technique constructor in a factory. This way you can get rid of the validate method, as it doesn't do match for all techniques except the z_score

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 17]
```diff
@@ -52,12 +72,36 @@ public void normalize(NormalizeScoresDTO normalizeScoresDTO) {
...
+        switch (combinationTechnique.techniqueName()) {
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:89]: I would prefer the class level constant Set with supported comb techniques, then you check with Set.contains

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 18]
```diff
@@ -52,12 +72,36 @@ public void normalize(NormalizeScoresDTO normalizeScoresDTO) {
...
+            case ArithmeticMeanScoreCombinationTechnique.TECHNIQUE_NAME:
+                // This is the supported technique, so we do nothing
+                break;
+            case GeometricMeanScoreCombinationTechnique.TECHNIQUE_NAME:
+                throw new IllegalArgumentException("Z Score does not support geometric mean combination technique");
+            case HarmonicMeanScoreCombinationTechnique.TECHNIQUE_NAME:
+                throw new IllegalArgumentException("Z Score does not support harmonic mean combination technique");
+            default:
+                throw new IllegalArgumentException(
+                    "Z Score does not support the provided combination technique {}: Supported technique is arithmetic_mean"
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:99]: change it to String.format(locale, message, param)

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/L2ScoreNormalizationTechnique.java [hunk 19]
```diff
@@ -69,6 +73,21 @@ public void normalize(final NormalizeScoresDTO normalizeScoresDTO) {
         }
     }
 
+    @Override
+    public void validateCombinationTechnique(ScoreCombinationTechnique combinationTechnique) {
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/L2ScoreNormalizationTechnique.java:77]: this method may be dropped, check my comment below

File: src/test/java/org/opensearch/neuralsearch/processor/ScoreNormalizationIT.java [hunk 20]
```diff
@@ -244,6 +248,95 @@ public void testMinMaxNorm_whenOneShardAndQueryMatches_thenSuccessful() {
...
+     *                 "combination": {
+     *                     "technique": "arithmetic_mean"
+     *                 }
+     *             }
+     *         }
+     *     ]
+     * }
+     */
+    @SneakyThrows
+    public void testZScoreNorm_whenOneShardAndQueryMatches_thenSuccessful() {
```
Comment (vibrantvarun) [inline src/test/java/org/opensearch/neuralsearch/processor/ScoreNormalizationIT.java:270]: Can we add one more test with sorting?

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 21]
```diff
@@ -0,0 +1,258 @@
...
+
+        float[] meanPerSubQuery = new float[numOfScores];
+        for (int i = 0; i < numOfScores; i++) {
+            meanPerSubQuery[i] = (float) statsPerSubquery[i].getMean();
+        }
+
+        return meanPerSubQuery;
+    }
+
+    private static float[] calculateStandardDeviationPerSubquery(final List<CompoundTopDocs> queryTopDocs, final int numOfScores) {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:211]:
```suggestion
    private static float[] calculateStandardDeviationPerSubquery(final List<CompoundTopDocs> queryTopDocs, final int numOfSubqueries) {
```

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+                for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
+                    statsPerSubquery[subQueryIndex].addValue(scoreDoc.score);
+                }
+            }
+        }
+
+        return statsPerSubquery;
+    }
+
+    private static float[] calculateMeanPerSubquery(final List<CompoundTopDocs> queryTopDocs, final int numOfScores) {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:200]:
```suggestion
    private static float[] calculateMeanPerSubquery(final List<CompoundTopDocs> queryTopDocs, final int numOfSubqueries) {
```

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+
+        float[] minPerSubQuery = new float[numOfScores];
+        for (int i = 0; i < numOfScores; i++) {
+            minPerSubQuery[i] = (float) statsPerSubquery[i].getMin();
+        }
+
+        return minPerSubQuery;
+    }
+
+    private static DescriptiveStatistics[] calculateStatsPerSubquery(final List<CompoundTopDocs> queryTopDocs, final int numOfScores) {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:178]:
```suggestion
    private static DescriptiveStatistics[] calculateStatsPerSubquery(final List<CompoundTopDocs> queryTopDocs, final int numOfSubqueries) {
```

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 24]
```diff
@@ -0,0 +1,258 @@
...
+
+        float[] maxPerSubQuery = new float[numOfScores];
+        for (int i = 0; i < numOfScores; i++) {
+            maxPerSubQuery[i] = (float) statsPerSubquery[i].getMax();
+        }
+
+        return maxPerSubQuery;
+    }
+
+    private static float[] calculateMinScorePerSubquery(final List<CompoundTopDocs> queryTopDocs, final int numOfScores) {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:167]:
```suggestion
    private static float[] calculateMinScorePerSubquery(final List<CompoundTopDocs> queryTopDocs, final int numOfSubqueries) {
```

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 25]
```diff
@@ -0,0 +1,258 @@
...
+        return queryTopDocs.stream()
+            .filter(Objects::nonNull)
+            .filter(topDocs -> !topDocs.getTopDocs().isEmpty())
+            .findAny()
+            .get()
+            .getTopDocs()
+            .size();
+    }
+
+    private static float[] calculateMaxScorePerSubquery(final List<CompoundTopDocs> queryTopDocs, final int numOfScores) {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:156]:
```suggestion
    private static float[] calculateMaxScorePerSubquery(final List<CompoundTopDocs> queryTopDocs, final int numOfSubqueries) {
```

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java [hunk 26]
```diff
@@ -0,0 +1,258 @@
...
+                        normalizedScore
+                    );
+                    scoreDoc.score = normalizedScore;
+                }
+            }
+        }
+        return getDocIdAtQueryForNormalization(normalizedScores, this);
+    }
+
+    private int getNumOfSubqueries(final List<CompoundTopDocs> queryTopDocs) {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java:146]: Can we move this method to a util class class like HybridQueryUtil or ProcessorUtil? This logic is used by almost every normalization technique. We can reuse the same in every class.

**Review by vibrantvarun — COMMENTED** [review]
Overall code looks good. Just few minor comments.

File: src/test/java/org/opensearch/neuralsearch/processor/ScoreNormalizationIT.java (hunk 20 above)
Comment (owaiskazi19) [reply src/test/java/org/opensearch/neuralsearch/processor/ScoreNormalizationIT.java:270]: Can you create a separate issue for this? Since such test needs to be added for all the techniques

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java (hunk 16 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java]: Used the DTO pattern

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationFactory.java [hunk 27]
```diff
@@ -42,4 +72,29 @@ public ScoreNormalizationTechnique createNormalization(final String technique, f
             .orElseThrow(() -> new IllegalArgumentException("provided normalization technique is not supported"))
             .apply(params);
     }
+
+    /**
+     * Validate normalization technique based on combination technique and other params that needs to be validated
+     * @param normalizationTechnique normalization technique to be validated
+     * @param validateNormalizationDTO data transfer object that contains combination technique and other params that needs to be validated
+     */
+    public void validateNormalizationTechnique(
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationFactory.java:81]: I kept the method generic here since we can add other validation parameters in the DTO

[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationFactory.java:81]: exceptions are thrown from each if cases and we can extend with other validations as well. If we just return a boolean here then we won't be able to handle the multiple cases of validation

File: src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombinationTechnique.java [hunk 28]
```diff
@@ -12,4 +12,6 @@ public interface ScoreCombinationTechnique {
      * @return combined score
      */
     float combine(final float[] scores);
+
+    String techniqueName();
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombinationTechnique.java:16]: please add javadoc comment, it's in public interface

File: src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java [hunk 29]
```diff
@@ -77,6 +78,13 @@ public SearchPhaseResultsProcessor create(
             Map<String, Object> combinationParams = readOptionalMap(NormalizationProcessor.TYPE, tag, combinationClause, PARAMETERS);
             scoreCombinationTechnique = scoreCombinationFactory.createCombination(combinationTechnique, combinationParams);
         }
+
+        TechniqueCompatibilityCheckDTO techniqueCompatibilityCheckDTO = TechniqueCompatibilityCheckDTO.builder()
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java:82]: If you are not going to rename the method to "isCompatible" then your old class name ValidateNormalizationDTO 
 fits better

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationFactory.java [hunk 30]
```diff
@@ -23,7 +30,32 @@ public class ScoreNormalizationFactory {
         L2ScoreNormalizationTechnique.TECHNIQUE_NAME,
         params -> new L2ScoreNormalizationTechnique(params, scoreNormalizationUtil),
         RRFNormalizationTechnique.TECHNIQUE_NAME,
-        params -> new RRFNormalizationTechnique(params, scoreNormalizationUtil)
+        params -> new RRFNormalizationTechnique(params, scoreNormalizationUtil),
+        ZScoreNormalizationTechnique.TECHNIQUE_NAME,
+        params -> new ZScoreNormalizationTechnique()
+    );
+
+    private final Map<String, Set<String>> combinationTechniqueForNormalizationTechniqueMap = Map.of(
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationFactory.java:38]: this can be a static constant

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationFactory.java [hunk 31]
```diff
@@ -40,4 +72,26 @@ public ScoreNormalizationTechnique createNormalization(final String technique, f
...
+
+    /**
+     * Validate normalization technique based on combination technique and other params that needs to be validated
+     * @param techniqueCompatibilityCheckDTO data transfer object that contains combination technique and other params that needs to be validated
+     */
+    public void validateNormalizationTechnique(TechniqueCompatibilityCheckDTO techniqueCompatibilityCheckDTO) {
+        ScoreNormalizationTechnique normalizationTechnique = techniqueCompatibilityCheckDTO.getScoreNormalizationTechnique();
+        Set<String> supportedTechniques = combinationTechniqueForNormalizationTechniqueMap.get(normalizationTechnique.techniqueName());
+
+        if (!supportedTechniques.contains(techniqueCompatibilityCheckDTO.getScoreCombinationTechnique().techniqueName())) {
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationFactory.java:84]:
```suggestion
        if (supportedTechniques.contains(techniqueCompatibilityCheckDTO.getScoreCombinationTechnique().techniqueName()) == false) {
```

**Comment by martin-gaievski** [comment]
Please open a documentation PR for this feature, you don't need to create complete content, at this stage it's mainly for tracking purposes

File: src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java (hunk 29 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java:82]: Changed the method to `isTechniquesCompatible`

File: src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationFactory.java [hunk 32]
```diff
@@ -17,13 +24,38 @@ public class ScoreNormalizationFactory {
 
     public static final ScoreNormalizationTechnique DEFAULT_METHOD = new MinMaxScoreNormalizationTechnique();
 
-    private final Map<String, Function<Map<String, Object>, ScoreNormalizationTechnique>> scoreNormalizationMethodsMap = Map.of(
+    private static final Map<String, Function<Map<String, Object>, ScoreNormalizationTechnique>> scoreNormalizationMethodsMap = Map.of(
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/normalization/ScoreNormalizationFactory.java:27]: this also needs to be renamed, once it became the static final constant, `SCORE_NORMALIZATION_METHODS`?

**Review by martin-gaievski — APPROVED** [review]
Overall looks good, thanks @owaiskazi19. Please address my comment, it's minor, and we're good.

**Comment by owaiskazi19** [comment]
> Please open a documentation PR for this feature, you don't need to create complete content, at this stage it's mainly for tracking purposes

https://github.com/opensearch-project/documentation-website/issues/9455

**Review by vibrantvarun — COMMENTED** [review]
Approved, but need to resolve merge conflicts.

**Review by vibrantvarun — APPROVED** [review]
